cmd/crank/completion/completion.go

```diff
@@ -0,0 +1,259 @@
+)
+
+// Predictors returns all supported predictors.
+func Predictors() map[string]complete.Predictor {
+	yamlPredictor := complete.PredictOr(
+		complete.PredictFiles("*.yml"),
+		complete.PredictFiles("*.yaml"),
+	)
+	return map[string]complete.Predictor{
+		"file":                   complete.PredictFiles("*"),
+		"xpkg_file":              complete.PredictFiles("*.xpkg"),
+		"yaml_file":              yamlPredictor,
+		"directory":              complete.PredictDirs("*"),
+		"file_or_directory":      complete.PredictOr(complete.PredictFiles("*"), complete.PredictDirs("*")),
+		"yaml_file_or_directory": complete.PredictOr(yamlPredictor, complete.PredictDirs("*")),
+		"namespace":              namespacePredictor(),
+		"context":                contextPredictor(),
+		"k8s_resource":           kubernetesResourcePredictor(),
+		"k8s_resource_name":      kubernetesResourceNamePredictor(),
+	}
+}
+
+func kubernetesResourcePredictor() complete.PredictFunc {
+	return func(a complete.Args) []string {
+		_, kubeconfig, err := kubernetesClient(parseConfigOverride(a))
+		if err != nil {
+			return nil
+		}
+
+		discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeconfig)
+		if err != nil {
+			return nil
+		}
+
+		resources, err := discoveryClient.ServerPreferredResources()
+		if err != nil {
+			return nil
+		}
+
+		if len(resources) == 0 {
+			return nil
+		}
+
+		var predictions []string
+		for _, apiResources := range resources {
+			for _, res := range apiResources.APIResources {
+				var resourceName string
+
+				switch {
+				case res.Group != "" && res.Version != "":
+					resourceName = fmt.Sprintf("%s.%s.%s", res.Name, res.Version, res.Group)
+				case res.Version != "":
+					resourceName = fmt.Sprintf("%s.%s", res.Name, res.Version)
+				default:
+					resourceName = res.Name
+				}
+
+				if strings.HasPrefix(resourceName, a.Last) {
+					predictions = append(predictions, resourceName)
+				}
+			}
+		}
+		return predictions
+	}
+}
+
+func kubernetesResourceNamePredictor() complete.PredictFunc {
+	return func(a complete.Args) []string {
+		client, kubeconfig, err := kubernetesClient(parseConfigOverride(a))
+		if err != nil {
+			return nil
+		}
+
+		discoveryClient, err := discovery.NewDiscoveryClientForConfig(kubeconfig)
+		if err != nil {
+			return nil
+		}
+
+		d := memory.NewMemCacheClient(discoveryClient)
+		rmapper := restmapper.NewShortcutExpander(restmapper.NewDeferredDiscoveryRESTMapper(d), d, nil)
+		mapping, err := mappingFor(rmapper, a.LastCompleted)
+		if err != nil {
+			return nil
+		}
+
+		// Get Resource object. Contains k8s resource and all its children, also as Resource.
+		u := &unstructured.UnstructuredList{}
+		u.SetGroupVersionKind(schema.GroupVersionKind{
+			Kind:    mapping.GroupVersionKind.Kind,
+			Group:   mapping.GroupVersionKind.Group,
+			Version: mapping.GroupVersionKind.Version,
+		})
+		err = client.List(context.Background(), u)
```

Comment:
when testing, I was surprised this seems to work for namespaced resources - we're not passing in any namespace information that I can see.

e.g., I type `crossplane beta trace acmedatabases.v1.acme.com `, which is a namespaced claim type, hit `<tab>`, and it then completes with the namespaced claim named `acme-db-prod`. 

Any idea how that may be working? 🤔

Reply:
Yes, in fact I was not passing a namespace to the client and I have only tested it with XRs, I believe. Now it should work for namespace-scoped resources in the same way as kubectl does. Taking the namespace from command line if specified, otherwise from the current kubeconfig.